### PR TITLE
Track user connections across clusters

### PR DESCRIPTION
**Pull request by rosstimothy**
Ensure that `proxy_ssh_sessions_total` is incremented in both the root and the leaf cluster if a session to a node in the leaf cluster is established through the root cluster. This now also prevents leaf clusters from terminating during a graceful shutdown if there is an active connection that was created from the root.

Fixes #32605.

### CODE REVIEW

File: lib/reversetunnel/agentpool.go [hunk 1]
```diff
@@ -531,6 +531,14 @@ func (p *AgentPool) transport(ctx context.Context, channel ssh.Channel, requests
 		proxySigner:          p.PROXYSigner,
 		forwardClientAddress: true,
 	}
+
+	if p.IsRemoteCluster {
+		if tracker, ok := p.ReverseTunnelServer.(interface{ TrackUserConnection() func() }); ok {
+			t.trackUserConnection = tracker.TrackUserConnection
+		}
```
Comment (fspmarshall) [inline lib/reversetunnel/agentpool.go:538]: I find this difficult to parse at a glance... anonymous interface declaring a function that returns a function inside an assignment inside an if statement is just one too many things I think.  I'd suggest declaring this interface separately.

Additionally, I think `TrackUserConnection` is the kind of function that really benefits from named return types:

```golang
type userConnectionTracker interface {
    TrackUserConnection() (cancel func())
}
```

intended usage pattern is much more obvious IMO.

File: lib/reversetunnel/srv.go [hunk 2]
```diff
@@ -1084,6 +1084,13 @@ func (s *server) rejectRequest(ch ssh.NewChannel, reason ssh.RejectionReason, ms
 	}
 }
 
+// TrackUserConnection tracks a user connection that should prevent
+// the server from being terminated if active. The returned function
+// should be called when the connection is terminated.
+func (s *server) TrackUserConnection() func() {
```
Comment (fspmarshall) [inline lib/reversetunnel/srv.go:1090]:
```suggestion
func (s *server) TrackUserConnection() (cancel func()) {
```

File: lib/reversetunnel/transport.go [hunk 3]
```diff
@@ -93,6 +94,8 @@ type transport struct {
 	// preventing users connecting to the proxy tunnel listener spoofing their address; but we are still able to
 	// correctly propagate client address in reverse tunnel agents of nodes/services.
 	forwardClientAddress bool
+
+	trackUserConnection func() func()
```
Comment (fspmarshall) [inline lib/reversetunnel/transport.go:98]:
```suggestion
	trackUserConnection func() (cancel func())
```

File: lib/sshutils/server.go [hunk 4]
```diff
@@ -441,6 +441,17 @@ func (s *Server) trackUserConnections(delta int32) int32 {
 	return atomic.AddInt32(&s.userConns, delta)
 }
 
+// TrackUserConnection tracks a user connection that should prevent
+// the server from being terminated if active. The returned function
+// should be called when the connection is terminated.
+func (s *Server) TrackUserConnection() func() {
```
Comment (fspmarshall) [inline lib/sshutils/server.go:447]:
```suggestion
func (s *Server) TrackUserConnection() (cancel func()) {
```

File: lib/reversetunnel/transport.go [hunk 5]
```diff
@@ -246,6 +250,11 @@ func (p *transport) start() {
 		// tunnel from the SSH node by dreq.ServerID. We'll need to forward
 		// dreq.Address as well.
 		directAddress = dreq.Address
+
+		if p.trackUserConnection != nil {
+			closed := p.trackUserConnection()
+			defer closed()
```
Comment (espadolini) [inline lib/reversetunnel/transport.go:256]: Why the additional variable?

```suggestion
			defer p.trackUserConnection()()
```

File: lib/reversetunnel/agentpool.go [hunk 6]
```diff
@@ -531,6 +531,23 @@ func (p *AgentPool) transport(ctx context.Context, channel ssh.Channel, requests
...
+
+	// If the AgentPool is being used for Proxy to Proxy communication between two clusters, then
+	// we check if the reverse tunnel server is capable of tracking user connections. This allows
+	// the leaf proxy to track sessions that are initiated via the root cluster. Without providing
+	// the user tracker the leaf cluster metrics will be incorrect and graceful shutdown will not
+	// wait for user sessions to be terminated prior to proceeding with the shutdown operation.
+	if p.IsRemoteCluster {
+		type userConnectionTracker interface {
+			TrackUserConnection() (closed func())
+		}
```
Comment (espadolini) [inline lib/reversetunnel/agentpool.go:543]: I tried adding `TrackUserConnection` to `reversetunnelclient.Server` and it seems to work. Why the additional interface?

File: lib/reversetunnel/transport.go [hunk 7]
```diff
@@ -93,6 +94,9 @@ type transport struct {
 	// preventing users connecting to the proxy tunnel listener spoofing their address; but we are still able to
 	// correctly propagate client address in reverse tunnel agents of nodes/services.
 	forwardClientAddress bool
+
+	// trackUserConnection is an optional mechanism used to count active user sessions.
+	trackUserConnection func() (closed func())
```
Comment (espadolini) [inline lib/reversetunnel/transport.go:99]: IMO

```suggestion
	trackUserConnection func() (release func())
```

**Review by espadolini — APPROVED** [review]

**Review by gabrielcorado — APPROVED** [review]

**Pull request merged**